File: app/src/test/java/tech/pauly/findapet/shared/MainViewModelTest.java
package tech.pauly.findapet.shared;

import android.support.annotation.NonNull;
import android.widget.ToggleButton;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import io.reactivex.Observable;
import io.reactivex.observers.TestObserver;
import tech.pauly.findapet.R;
import tech.pauly.findapet.data.models.AnimalType;
import tech.pauly.findapet.discover.DiscoverFragment;
import tech.pauly.findapet.favorites.FavoritesFragment;
import tech.pauly.findapet.settings.SettingsFragment;
import tech.pauly.findapet.shared.datastore.DiscoverAnimalTypeUseCase;
import tech.pauly.findapet.shared.datastore.DiscoverToolbarTitleUseCase;
import tech.pauly.findapet.shared.datastore.TransientDataStore;
import tech.pauly.findapet.shared.datastore.UseCase;
import tech.pauly.findapet.shared.events.BaseViewEvent;
import tech.pauly.findapet.shared.events.FragmentEvent;
import tech.pauly.findapet.shared.events.ViewEventBus;
import tech.pauly.findapet.shelters.SheltersFragment;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.clearInvocations;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class MainViewModelTest {

    @Mock
    private ViewEventBus eventBus;

    @Mock
    private TransientDataStore dataStore;

    @Mock
    private DiscoverToolbarTitleUseCase useCase;

    private MainViewModel subject;
    private TestObserver<MainViewModel.ExpandingLayoutEvent> expandingLayoutObserver = new TestObserver<>();
    private TestObserver<Boolean> drawerCloseObserver = new TestObserver<>();

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        when(useCase.getTitle()).thenReturn(R.string.tab_cat);
        when(dataStore.observeAndGetUseCase(any())).thenReturn(Observable.just(useCase));
        subject = new MainViewModel(eventBus, dataStore);
        subject.getExpandingLayoutSubject().subscribe(expandingLayoutObserver);
        subject.getDrawerSubject().subscribe(drawerCloseObserver);
    }

    @Test
    public void subscribeToDataStore_getToolbarTileUseCase_updateToolbarTitle() {
        subject.subscribeToDataStore();

        verify(dataStore).observeAndGetUseCase(DiscoverToolbarTitleUseCase.class);
        assertThat(subject.toolbarTitle.get()).isEqualTo(R.string.tab_cat);
    }

    @Test
    public void firstLaunch_launchesCatAnimalType() {
        subject.clickFirstAnimal();
        subject.subscribeToDataStore();

        verify(dataStore).save(new DiscoverAnimalTypeUseCase(AnimalType.CAT));
        verify(eventBus).send(new FragmentEvent(subject, DiscoverFragment.class, R.id.fragment_content));
    }

    @Test
    public void notFirstLaunch_doNotLaunchFragment() {
        subject.clickFirstAnimal();
        subject.subscribeToDataStore();
        clearInvocations(dataStore);
        clearInvocations(eventBus);

        subject.subscribeToDataStore();

        verify(dataStore, never()).save(any(UseCase.class));
        verify(eventBus, never()).send(any(BaseViewEvent.class));
    }

    @Test
    public void clickAnimalType_launchDiscoverFragmentForAnimalTypeAndSetsCurrentAnimalType() {
        subject.clickAnimalType(AnimalType.BARNYARD);

        assertThat(subject.currentAnimalType.get()).isEqualTo(AnimalType.BARNYARD);
        verify(dataStore).save(new DiscoverAnimalTypeUseCase(AnimalType.BARNYARD));
        verify(eventBus).send(new FragmentEvent(subject, DiscoverFragment.class, R.id.fragment_content));
        drawerCloseObserver.assertValue(true);
    }

    @Test
    public void clickStandardButton_buttonAlreadyChecked_stayChecked() {
        ToggleButton button = mock(ToggleButton.class);
        when(button.isChecked()).thenReturn(false);

        subject.clickShelters(button);

        verify(button).setChecked(true);
        drawerCloseObserver.assertValue(true);
        assertThat(subject.currentAnimalType.get()).isNull();
    }

    @Test
    public void clickStandardButton_buttonNotChecked_collapseDiscoverViewAndSetCurrentButtonAndResetCurrentAnimal() {
        ToggleButton button = mockButton();

        subject.clickShelters(button);

        expandingLayoutObserver.assertValue(MainViewModel.ExpandingLayoutEvent.COLLAPSE);
        assertThat(subject.currentButton.get()).isEqualTo(1);
        verify(button, never()).setChecked(true);
        drawerCloseObserver.assertValue(true);
        assertThat(subject.currentAnimalType.get()).isNull();
    }

    @Test
    public void clickShelters_launchesSheltersFragment() {
        subject.clickShelters(mockButton());

        verify(eventBus).send(new FragmentEvent(subject, SheltersFragment.class, R.id.fragment_content));
    }

    @Test
    public void clickFavoritesButton_launchesFavoritesFragment() {
        subject.clickFavorites(mockButton());

        verify(eventBus).send(new FragmentEvent(subject, FavoritesFragment.class, R.id.fragment_content));
    }

    @Test
    public void clickSettingsButton_launchesSettingsFragment() {
        subject.clickSettings(mockButton());

        verify(eventBus).send(new FragmentEvent(subject, SettingsFragment.class, R.id.fragment_content));
    }

    @NonNull
    private ToggleButton mockButton() {
        ToggleButton button = mock(ToggleButton.class);
        when(button.isChecked()).thenReturn(true);
        when(button.getId()).thenReturn(1);
        return button;
    }
}

File: app-old/src/test/java/tech/pauly/old/shared/MainViewModelTest.kt
package tech.pauly.old.shared

import android.widget.ToggleButton
import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import io.reactivex.Observable
import io.reactivex.observers.TestObserver
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito.*
import org.mockito.MockitoAnnotations
import tech.pauly.old.R
import tech.pauly.old.data.models.AnimalType
import tech.pauly.old.discover.DiscoverFragment
import tech.pauly.old.favorites.FavoritesFragment
import tech.pauly.old.settings.SettingsFragment
import tech.pauly.old.shared.datastore.DiscoverAnimalTypeUseCase
import tech.pauly.old.shared.datastore.DiscoverToolbarTitleUseCase
import tech.pauly.old.shared.datastore.TransientDataStore
import tech.pauly.old.shared.events.FragmentEvent
import tech.pauly.old.shared.events.ViewEventBus
import tech.pauly.old.shelters.SheltersFragment

class MainViewModelTest {

    private val eventBus: ViewEventBus = mock()
    private val dataStore: TransientDataStore = mock()
    private val useCase: DiscoverToolbarTitleUseCase = mock()

    private lateinit var subject: MainViewModel
    private val expandingLayoutObserver = TestObserver<MainViewModel.ExpandingLayoutEvent>()
    private val drawerCloseObserver = TestObserver<Boolean>()

    @Before
    fun setup() {
        MockitoAnnotations.initMocks(this)
        whenever(useCase.title).thenReturn(R.string.tab_cat)
        whenever(dataStore.observeAndGetUseCase(DiscoverToolbarTitleUseCase::class)).thenReturn(Observable.just(useCase))
        subject = MainViewModel(eventBus, dataStore)
        subject.expandingLayoutSubject.subscribe(expandingLayoutObserver)
        subject.drawerSubject.subscribe(drawerCloseObserver)
    }

    @Test
    fun subscribeToDataStore_getToolbarTileUseCase_updateToolbarTitle() {
        subject.subscribeToDataStore()

        verify(dataStore).observeAndGetUseCase(DiscoverToolbarTitleUseCase::class)
        assertThat(subject.toolbarTitle.get()).isEqualTo(R.string.tab_cat)
    }

    @Test
    fun firstLaunch_launchesCatAnimalType() {
        subject.clickFirstAnimal()
        subject.subscribeToDataStore()

        verify(dataStore) += DiscoverAnimalTypeUseCase(AnimalType.CAT)
        verify(eventBus) += FragmentEvent(subject, DiscoverFragment::class, R.id.fragment_content)
    }

    @Test
    fun notFirstLaunch_doNotLaunchFragment() {
        subject.clickFirstAnimal()
        subject.subscribeToDataStore()
        clearInvocations(dataStore)
        clearInvocations(eventBus)

        subject.subscribeToDataStore()

        verify(dataStore, never()) += any()
        verify(eventBus, never()) += any()
    }

    @Test
    fun clickAnimalType_launchDiscoverFragmentForAnimalTypeAndSetsCurrentAnimalType() {
        subject.clickAnimalType(AnimalType.BARNYARD)

        assertThat(subject.currentAnimalType.get()).isEqualTo(AnimalType.BARNYARD)
        verify(dataStore) += DiscoverAnimalTypeUseCase(AnimalType.BARNYARD)
        verify(eventBus) += FragmentEvent(subject, DiscoverFragment::class, R.id.fragment_content)
        drawerCloseObserver.assertValue(true)
    }

    @Test
    fun clickStandardButton_buttonAlreadyChecked_stayChecked() {
        val button: ToggleButton = mock()
        whenever(button.isChecked).thenReturn(false)

        subject.clickShelters(button)

        verify(button).isChecked = true
        drawerCloseObserver.assertValue(true)
        assertThat(subject.currentAnimalType.get()).isNull()
    }

    @Test
    fun clickStandardButton_buttonNotChecked_collapseDiscoverViewAndSetCurrentButtonAndResetCurrentAnimal() {
        val button = mockButton()

        subject.clickShelters(button)

        expandingLayoutObserver.assertValue(MainViewModel.ExpandingLayoutEvent.COLLAPSE)
        assertThat(subject.currentButton.get()).isEqualTo(1)
        verify(button, never()).isChecked = true
        drawerCloseObserver.assertValue(true)
        assertThat(subject.currentAnimalType.get()).isNull()
    }

    @Test
    fun clickShelters_launchesSheltersFragment() {
        subject.clickShelters(mockButton())

        verify(eventBus) += FragmentEvent(subject, SheltersFragment::class, R.id.fragment_content)
    }

    @Test
    fun clickFavoritesButton_launchesFavoritesFragment() {
        subject.clickFavorites(mockButton())

        verify(eventBus) += FragmentEvent(subject, FavoritesFragment::class, R.id.fragment_content)
    }

    @Test
    fun clickSettingsButton_launchesSettingsFragment() {
        subject.clickSettings(mockButton())

        verify(eventBus) += FragmentEvent(subject, SettingsFragment::class, R.id.fragment_content)
    }

    private fun mockButton(): ToggleButton {
        val button: ToggleButton = mock()
        whenever(button.isChecked).thenReturn(true)
        whenever(button.id).thenReturn(1)
        return button
    }
}